### DISCUSSION TITLE
How to configure support for reading vs. writing vs both?

### DISCUSSION THREAD

**Discussion by axxel** [zxing-cpp/zxing-cpp#744] [Polls]
There always has been the option to disable the build of the readers or the writers via cmake `-DBUILD_READERS=OFF` or `-DBUILD_WRITERS=OFF`. This then removed the API and the implementation from the project. The (only) reason to do that would be to save on binary size. With the inclusion (either as addition or as replacement) of the `libzint` project this incentive becomes a little bigger because this adds 500k writer to the 1000K reader and if the old writers are removed, it is still a net addition of 200K.

With all the wrappers the configuration of this becomes a bit tricky. My question is now how to do that most effectively and how important is this anyway?

To make life easier with respect to the build configuration, I'm currently leaning toward the last of the options below, meaning the API is always complete but parts of it will always fail. This would not produce an absolute minimal build and also require a bit more source code in the form of 'dummy' implementations but it would do away with the requirement to have the wrappers cope with a configuration dependent API. It would make sense then to add a simple query function that can be used to determine what capabilities the implementation has. This could actually be used to get a detailed answer as to which symbologies are supported, something like `BarcodeFormats SupportedBarcodeFormats(Operation op = Operation::ReadAndCreate)`, where `Operation` would be an enum: `Read, Create, ReadAndCreate`. That in turn would allow to potentially select individual symbologies to support, like a QRCode-only version. Might come in very handy in #715 type situations.

If you choose anything other than the first, I'd be happy to read some comments including what interface/wrapper you are interested in.

**Comment by pranavtriguna** [comment]
Hi again !
Since I'm the only one who seems to care about this, my take on this is that it is great to have multiple build options available to customize, since they are fundamentally different operations in terms of real-world applicability.
It helps to be able to build separately, with the benefit of performance and/or code size. This also expands the potential pool of HW platforms this can be used for.
Of course, this is for people who care about things like strip down code and cycle counts :)

**Reply by axxel** [reply]
To clarify: I will not remove the option to build without read/write support. The question is: how do I do that? I prefer the 4th option right now because it limits the effect to cmake/c++ while all wrappers can be unencumbered.

The advantage of option 3 over 4 would be an extra tiny bit of saved space, I would currently guess in the order of 1%.

**Reply by pranavtriguna** [reply]
Aah, as long as there's an option to not build what is not required, any approach is fine.
Disadvantage of option 3 is we cannot "re-enable" features later if we want. Hence option 4 has most diverse set of applications.

**Comment by axxel** [comment]
I'm closing this discussion because the matter has been settled. See `ZXING_READERS`, `ZXING_WRITERS`, `ZXING_ENABLE_...` and `BarcodeFormat::list()`.
